libcnb-cargo/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+    }
+
+    let exit_status = Command::new("cargo")
+        .args(cargo_args)
+        .envs(
+            cross_compile::cross_compile_env(target_triple.as_ref())
+                .map_err(BuildError::CrossCompileError)?,
+        )
+        .current_dir(&project_path)
+        .spawn()
+        .and_then(|mut child| child.wait())
+        .map_err(BuildError::IoError)?;
+
+    if exit_status.success() {
+        let binary_path = cargo_metadata
+            .target_directory
+            .join(target_triple.as_ref())
+            .join(match cargo_profile {
+                CargoProfile::Dev => "debug",
+                CargoProfile::Release => "release",
+            })
+            .join(&target.name)
+            .into_std_path_buf();
+
+        Ok(binary_path)
+    } else {
+        Err(BuildError::UnexpectedExitStatus(exit_status))
+    }
+}
+
+#[derive(Debug)]
+pub enum BuildError {
+    IoError(std::io::Error),
+    UnexpectedExitStatus(ExitStatus),
+    CrossCompileError(CrossCompileError),
+    NoTargetsFound,
+    MultipleTargetsFound,
+    MetadataError(cargo_metadata::Error),
+    CouldNotFindRootPackage,
+}
+
+#[derive(Copy, Clone)]
+pub enum CargoProfile {
+    Dev,
+    Release,
+}
+
+/// Reads buildpack data from the given project path.
+///
+/// # Errors
+///
+/// Will return `Err` if the buildpack data could not be read successfully.
+pub fn read_buildpack_data(
+    project_path: impl AsRef<Path>,
+) -> Result<BuildpackData<Option<toml::Value>>, BuildpackDataError> {
+    let buildpack_toml_path = project_path.as_ref().join("buildpack.toml");
+
+    fs::read_to_string(&buildpack_toml_path)
+        .map_err(BuildpackDataError::IoError)
+        .and_then(|file_contents| {
+            toml::from_str(&file_contents).map_err(BuildpackDataError::DeserializationError)
+        })
+        .map(|buildpack_toml| BuildpackData {
+            buildpack_toml_path,
+            buildpack_toml,
+        })
+}
+
+#[derive(Debug)]
+pub enum BuildpackDataError {
+    IoError(std::io::Error),
+    DeserializationError(toml::de::Error),
+}
+
+pub struct BuildpackData<BM> {
+    pub buildpack_toml_path: PathBuf,
+    pub buildpack_toml: BuildpackToml<BM>,
+}
+
+/// Creates a buildpack directory and copies all buildpack assets to it.
+///
+/// Assembly of the directory follows the constraints set by the libcnb framework. For example,
+/// the buildpack binary is only copied once and symlinks are used to refer to it when the CNB
+/// spec requires different file(name)s.
+///
+/// This function will not validate if the buildpack descriptor at the given path is valid and will
+/// use it as-is.
+///
+/// # Errors
+///
+/// Will return `Err` if the buildpack directory could not be assembled.
+pub fn assemble_buildpack_directory(
+    destination_path: impl AsRef<Path>,
+    buildpack_toml_path: impl AsRef<Path>,
+    buildpack_binary_path: impl AsRef<Path>,
+) -> std::io::Result<()> {
+    fs::create_dir_all(destination_path.as_ref())?;
```

Comment:
I think we should remove the directory if it already exists, since making users delete it before re-running `cargo libcnb package` isn't ideal for UX.

My understanding is that the concern is over accidentally deleting a user-provided `output-path` (eg if they specify `~/`)?

Possible options:
1. Remove `output-path` for now and use only the default paths, which are safe to delete.
2. Change `output-path` to instead be a containing directory, rather than the actual output directory. For example `--output-path ~/foo` would result in the deletion/creation of `~/foo/my_buildpack/`, which is safe to delete.
3. Leave `output-path` as is for now, and perform the directory deletion only iff `output-path` is not used.

It's worth noting that when we add workspace support to `cargo libcnb package` in a future PR, we're going to have to do (2) for workspaces at the least. As such, `output-path` is more complex than it seems at first glance, and since this PR is an MVP I think it may be best to just drop that feature for now (and add it back in a later PR).

Reply:
I still believe that implicitly deleting is dangerous in any case, even if we don't allow custom output directories. The situation was a bit easier when we had tarballs instead since they were atomic.

I agree that output-path is tricky in the workspace scenario and removal of that feature might be the best option at this point. Since that also makes deletion much less dangerous, I'm happy to implement that as well. I'll add a comment when I'm done with that.

Reply:
:+1: on just removing the configuration for now.

Another suggestion is we can just use a tmpdir for each build vs a static location and just output where it's built. This way we know it's safe to prune/delete/recreate.

Reply:
Resolved with 7f35903 and 80d2cc2